v2/apiserver/internal/core/events.go

```diff
@@ -537,35 +537,41 @@ func (e *eventsService) CancelMany(
 		)
 	}
 
-	events, err := e.eventsStore.CancelMany(ctx, selector)
+	eventCh, err := e.eventsStore.CancelMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error canceling events in store")
 	}
 
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker, more
-	// efficient way to do this?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
-						err,
-						"error deleting event %q worker and jobs from the substrate",
-						event.ID,
-					),
-				)
+	// When an event comes in via the corresponding channel,
+	// kick off a go routine to execute cleanup.
+	for {
```

Comment:
I think this whole loop should be in its own goroutine, otherwise, we could be here a while for a large number of cancelations. We may be launching a goroutine for _each_ cancelation (more on that below), but for say a million cancelations, we'd still be waiting around while we drain 1m events from the channel and launch those goroutines. That can be time consuming in and of itself.